### ISSUE TITLE
Ceph FS support 

### ISSUE THREAD

**Issue by twelfthdoctor** [lxc/lxd#4822]
Hello. LXD 3.2 supports Ceph RBD as a storage backend. It will be very good to add a support for pure Ceph (CephFS).
Unfortunately, there is no support of such FS, so I have to mount my CephFS to /mnt/ceph
`sudo mount -t ceph 192.168.0.1:6789:/ /mnt/ceph -o name=admin,secretfile=/etc/ceph/admin.secret`
 and then create an additional dir storage pool
`lxc storage create cephfs dir source=/mnt/ceph`

**Comment by stgraber** [comment]
This would be a full storage driver on its own, so possibly quite a bit of work.
We'd need to see quite a bit more interest for it and ideally confirmation that running production workloads in a setup like the one you did with the `dir` backend does work properly.

(Confirming that cephfs is fully POSIX compliant, including ACLs support, xattrs, fscaps, ...)

**Comment by twelfthdoctor** [comment]
@stgraber if I set some quotas outside the container 
`setfattr -n ceph.quota.max_bytes -v 100000000 /mnt/ceph/vol1` will this quotas work inside a container? Note that CephFS quotas are not supported by Linux kernel (it was patched 5 Jan 2018).

**Comment by stgraber** [comment]
Did you some testing here after I enabled it on our test cluster.

Creating a new filesystem:
```
root@ceph01:~# ceph osd pool create testcephfs_data 8
pool 'testcephfs_data' created
root@ceph01:~# ceph osd pool create testcephfs_metadata 8
pool 'testcephfs_metadata' created
root@ceph01:~# ceph fs new testcephfs testcephfs_metadata testcephfs_data
new fs with metadata pool 254289 and data pool 254288
```

Mounting it (with FUSE in this case but we'll want to use the kernel driver instead):
```
root@ceph01:~# ceph-fuse --client_mds_namespace testcephfs /mnt/
2019-04-10 17:46:45.244500 7f6d7fdbd080 -1 init, newargv = 0x56333996fec0 newargc=9ceph-fuse[8235
]: starting ceph client
ceph-fuse[8235]: starting fuse
root@ceph01:~# ls /mnt/
root@ceph01:~# df -h
Filesystem                      Size  Used Avail Use% Mounted on
internal/lxd/containers/ceph01  808G  745M  808G   1% /
none                            492K     0  492K   0% /dev
udev                            126G     0  126G   0% /dev/tty
tmpfs                           100K     0  100K   0% /dev/lxd
tmpfs                            32G  5.1G   27G  16% /var/lib/ceph/journal
/dev/loop9                       30G  198M   30G   1% /var/lib/ceph/osd/ceph-0
tmpfs                           100K     0  100K   0% /dev/.lxd-mounts
tmpfs                           126G     0  126G   0% /dev/shm
tmpfs                           126G  8.2M  126G   1% /run
tmpfs                           5.0M     0  5.0M   0% /run/lock
tmpfs                           126G     0  126G   0% /sys/fs/cgroup
ceph-fuse                        29G     0   29G   0% /mnt
root@ceph01:~# touch /mnt/blah
root@ceph01:~# fusermount -u /mnt/
root@ceph01:~# ceph-fuse --client_mds_namespace testcephfs /mnt/
2019-04-10 17:47:02.392413 7fc6037f4080 -1 init, newargv = 0x558bdb9cdec0 newargc=9ceph-fuse[8270]: starting ceph client

ceph-fuse[8270]: starting fuse
root@ceph01:~# ls /mnt/
blah
root@ceph01:~# fusermount -u /mnt/
```

When snapshots are enabled, they can be used with:
```
root@ceph01:/mnt# mkdir .snap/blah
root@ceph01:/mnt# ls -lh .snap/blah
total 512
drwxr-xr-x 1 root root 0 Apr 10 17:57 1
-rw-r--r-- 1 root root 0 Apr 10 17:46 blah
```

Snapshots are immutable and can be restored by copying files from them.

**Comment by stgraber** [comment]
My suggestion for LXD's implementation is:
 - A new `cephfs` storage type with the following config options:
  - ceph.cluster_name
  - ceph.osd.pg_num
  - ceph.fs.name
  - ceph.user.name
  - rsync.bwlimit
  - source
  - volume.size
 - Support the following configurations:
  - Existing empty full cephfs: lxc storage create cephfs cephfs source=foo
  - Existing empty directory in cephfs: lxc storage create cephfs cephfs source=foo/bar
  - Non-existing cephfs: lxc storage create cephfs cephfs source=foo
  - Non-existing cephfs directory: lxc storage create cephfs cephfs source=foo/bar
 - Restrict to only custom volumes at this point
 - Use cephfs snapshot when available (fail snapshot creation if not)
 - Mount only the subset of the cephfs tree that's dedicated to LXD

So this will normally result in one cephfs mount per LXD daemon with LXD assuming full control over that part of the cephfs tree, LXD will then treat that mostly as it would the `dir` or `btrfs` backends with custom handling for:
 - Snapshot and snapshot restore
 - Setting quotas on volumes

As far as I can tell, there is no optimized network transfer for migration so we should just be using good old rsync for that.

**Comment by stgraber** [comment]
@monstermunchkin our CEPH cluster supports cephfs so we're just blocked on the internal storage rework before we can get this one done.

**Comment by stgraber** [comment]
So considering how difficult it is to delete a cephfs, I think we should change our plan a bit and instead only allow using an existing cephfs or directory on an existing cephfs.

Keeping the creation of the cephfs itself out of scope, at least until ceph handles multiple cephfs and deletion of existing ones better.
